### ISSUE TITLE
terragrunt >v0.44.0 increases runtime by at least 2x

### ISSUE THREAD

**Issue by tjstansell** [gruntwork-io/terragrunt#2480]
I don't know what's going on, but our docker image picked up the latest `v0.44.4` version this past weekend (was `v0.44.0` prior) and now a bunch of our highly parallel repos are taking 2x longer to do the same daily plan they've done for months.  We have one repo that normally takes 13.5 minutes and includes 511 separate terragrunt directories.  This is now taking 31 minutes to run.  The only thing I can see that's different in our image is this terragrunt upgrade.  Another of our repos normally does a daily apply across its 187 terragrunt directories in about 6.5 minutes.  It now takes 25 minutes.

I suspect this has something to do with the changes in #2467 (v0.44.1).  I don't know what you're doing, but I suspect your introducing a level of serialization related to hooks that didn't used to be there.  Any ideas on how to figure out what's going on here and where the increase runtimes are coming from?

**Comment by tjstansell** [comment]
FYI, just to be sure, I pinned out docker image to v0.44.0 and re-ran the daily apply and the runtime went from 25 minutes to 4.5 minutes.
![image](https://user-images.githubusercontent.com/5217454/223219043-cd50e825-803c-46d5-9d36-02c96ebd090b.png)

**Comment by levkoburburas** [comment]
@tjstansell, could you please provide some configurations to reproduce this issue?

**Comment by tjstansell** [comment]
OK.  Here's a simple script that'll generate a test setup ...
```bash
#!/bin/bash

init() {
    echo "$(pwd): init"
    cat >terragrunt.hcl <<'EOM'
include "root" {
  path = find_in_parent_folders("terragrunt-test.hcl")
}
EOM
    cat >main.tf <<'EOM'
EOM
}

cat >terragrunt-test.hcl <<'EOM'
terraform {
  # so our terragrunt logging shows more progress
  before_hook "starting" {
    commands = ["init", "plan", "apply"]
    execute  = ["true"]
  }

  before_hook "setup-module-cache" {
    commands = ["init"]
    execute = ["sleep", "2"]
  }

  before_hook "pre-apply-plan" {
    commands = ["apply"]
    execute = ["sleep", "5"]
  }

  after_hook "plan-convert" {
    commands = ["plan"]
    execute = ["sleep", "1"]
  }

  after_hook "complete" {
    commands     = ["init", "plan", "apply"]
    execute      = ["true"]
    run_on_error = true
  }
}
EOM

for i in $(seq 10); do
    mkdir -p acct$i
    pushd acct$i >/dev/null
    init
    for j in $(seq 4); do
        mkdir -p region$j
        pushd region$j >/dev/null
        init
        popd >/dev/null
    done
    popd >/dev/null
done
```

And then a couple of test runs to show the difference ... 
```
root@39df629f66dd:/app/examples/test# terragrunt --version
terragrunt version v0.44.0
root@39df629f66dd:/app/examples/test# time terragrunt run-all plan --terragrunt-log-level=debug >out.v0.44.0 2>&1

real	0m11.994s
user	0m2.357s
sys	0m2.534s
root@39df629f66dd:/app/examples/test# ./terragrunt-v0.44.4 --version
terragrunt version v0.44.4
root@39df629f66dd:/app/examples/test# time ./terragrunt-v0.44.4 run-all plan --terragrunt-log-level=debug >out.v0.44.4 2>&1

real	0m52.601s
user	0m2.545s
sys	0m2.807s
```

**Comment by tjstansell** [comment]
And actually, to be slightly more accurate for our normal setup, each regional deployment depends on the account, so if you update the setup script with this instead:
```bash
init() {
    echo "$(pwd): init"
    cat >terragrunt.hcl <<'EOM'
include "root" {
  path = find_in_parent_folders("terragrunt-test.hcl")
}
EOM
    if [ -f ../terragrunt.hcl ] ; then
        cat >>terragrunt.hcl <<EOM
dependencies {
  paths = ["../"]
}
EOM
    fi
    cat >main.tf <<'EOM'
EOM
}
```
Then run that, you'll see it breaks it into the accounts first, then all the regions ... and here's an updated test run, which doesn't seem to change the results much ...
```
root@39df629f66dd:/app/examples/test# terragrunt --version; time terragrunt run-all plan >/dev/null 2>&1
terragrunt version v0.44.0

real	0m11.697s
user	0m2.444s
sys	0m2.619s
root@39df629f66dd:/app/examples/test# ./terragrunt-v0.44.4 --version; time ./terragrunt-v0.44.4 run-all plan >/dev/null 2>&1
terragrunt version v0.44.4

real	0m51.860s
user	0m3.193s
sys	0m3.883s
```

**Comment by levkohimins** [comment]
@tjstansell, many thanks for such detailed analysis! My colleague seems to fix it in [v0.45.15](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.45.15) release. If the issue still persists, feel free to let us know.
